libraries/botbuilder-ai/tests/qnaMakerDialog.test.js

```diff
@@ -242,4 +251,201 @@ describe('QnAMakerDialog', function () {
+                return dm.onTurn(turnContext);
+            });
+
+            await adapter
+                .send('QnaMaker_TopNAnswer.json')
+                .assertReply((reply) => {
+                    strictEqual(reply.type, ActivityTypes.Message);
+                    strictEqual(reply.suggestedActions.actions.length, 3);
+                    strictEqual(reply.text, cardNoMatchText);
+                })
+                .startTest();
+        });
+
+        it('should error if suggestionsActivityFactory is passed in with falsy cardNoMatchText', () => {
+            const kbId = 'dummyKbId';
+            const endpointKey = 'dummyEndpointKey';
+            throws(
+                () => new QnAMakerDialog(kbId, endpointKey, HOSTNAME, undefined, undefined, (_) => {}, undefined),
+                new TypeError('cardNoMatchText is required when using the suggestionsActivityFactory.')
+            );
+        });
+
+        it('should error if suggestionsActivityFactory returns void', async () => {
+            const kbId = 'dummyKbId';
+            const endpointKey = 'dummyEndpointKey';
+            const convoState = new ConversationState(new MemoryStorage());
+            const dm = new DialogManager();
+            dm.conversationState = convoState;
+            const cardNoMatchText = 'Not helpful.';
+            const qnaDialog = new QnAMakerDialog(
+                kbId,
+                endpointKey,
+                HOSTNAME,
+                undefined,
+                undefined,
+                (suggestionsList, noMatchingQuestionsText) => {
+                    ok(suggestionsList);
+                    strictEqual(suggestionsList.length, 3);
+                    strictEqual(noMatchingQuestionsText, cardNoMatchText);
+                },
+                cardNoMatchText
+            );
+
+            dm.rootDialog = qnaDialog;
+            const adapter = new TestAdapter((turnContext) => {
+                return dm.onTurn(turnContext);
+            });
+
+            await adapter
+                .send('QnaMaker_TopNAnswer.json')
+                .catch((reason) => {
+                    ok(reason);
+                    strictEqual(reason.name, 'DialogContextError');
+                    strictEqual(reason.message, 'Activity with Active Learning suggestions was not generated.');
+                })
+                .startTest();
+        });
+
+        it('should error if QnACardBuilder.getSuggestionsCard returns void', async () => {
+            const kbId = 'dummyKbId';
+            const endpointKey = 'dummyEndpointKey';
+            const convoState = new ConversationState(new MemoryStorage());
+            const dm = new DialogManager();
+            dm.conversationState = convoState;
+            const suggestionsCardTitle = 'Card Title';
+            const cardNoMatchText = 'Not helpful.';
+            QnACardBuilder.getSuggestionsCard = (suggestionsList, cardTitle, noMatchingQuestionsText) => {
```

Comment:
Sinon sandboxing with mocks/stubs is a bit cleaner than doing the manual `beforeEach`/`afterEach` function patching. This can be updated later, but this is what the code ends up looking like:
```js
sandbox.mock(object).expects('methodName').withArgs(arg1, arg2).once().throws(new Error());
```
you do need a sinon sandbox which, for mocha, is created using beforeEach/afterEach.

Reply:
Note: I would save this for a future PR.

Reply:
Fixed via https://github.com/microsoft/botbuilder-js/pull/3505